src/include/optimizer/statistics/stats_storage.h

```diff
@@ -61,52 +100,72 @@ namespace noisepage::optimizer {
  * Manages all the existing table stats objects. Stores them in an
  * unordered map and keeps track of them using their database and table oids. Can
  * add, update, or delete table stats objects from the storage map.
+ *
+ * TODO(Joe Koshakow) Currently there is no cache eviction policy, which means that stats storage will grow forever. If
+ *  this becomes an issue we can implement a max number of columns to store with some eviction policy
  */
 class StatsStorage {
  public:
   /**
-   * Using given database and table ids,
-   * select a pointer to the TableStats objects in the table stats storage map.
+   * Returns a reference of a TableStats object for a specific table, and an acquired read lock on the entire Stats
+   * Storage
+   *
    * @param database_id - oid of database
    * @param table_id - oid of table
-   * @return pointer to a TableStats object
+   * @param accessor - catalog accessor
+   * @return reference to a TableStats object, latch for that TableStats object, and an acquired shared latch on
+   * StatsStorage
```

Comment:
Maybe document that the consumer of `empty_nullable_table` needs to take a shared latch when reading from `StatsStorageReference` and explain how to access that latch.

Reply:
@malin1993ml What do you think of this idea, I change `StatsStorageReference` to look something like this
```C++
struct StatsStorageReference {
 /** Table Statistics */
  const TableStats &table_stats_;
  /** Acquired Shared Latch on Table Stats */
  common::SharedLatchGuard table_stats_shared_latch_;
  /** Acquired Shared Latch on Stats Storage */
  common::SharedLatchGuard stats_storage_shared_latch_;
};
```
That when we return a table stats object it not only comes with an acquired latch on stats storage but it also comes with an acquired latch on the table stats itself.

I think the pro is that the consumer of this method doesn't need to do any additional locking which is probably less error-prone. The con is that if the consumer isn't using the table stats right away then they are unnecessarily holding a lock on it. Right now the consumer always consumes the table stats right away. Also it would probably be bad practice to get table stats from stats storage until right before you need it because you need a lock on stats storage anyway.

Also if you have a better name than `StatsStorageReference` please let me know.

Reply:
I actually like this idea a bit better than the original one. Please just document that in `GetTableStats` so that other people would notice the implication. Maybe call this `LockedTableStatsReference`.

Reply:
I'm going to push this change for now and in case there's an issue with it or you don't like it, then I'll revert it.

Reply:
I think the changes look good! Thanks! I do think it would be a little bit better to rename `StatsStorageReference` to `LatchedTableStatsReference`, since that's what it really is. Also, it might be better to rename `StatsStorageKey` to `TableStatsKey` and `StatsStorageValue` to `TableStatsValue`. Again, those are what they really are.

Reply:
Sorry, I did not see your last two messages until now. We must have been typing at the same time. I've updated everything with your suggestions.